ledger/complete/wal/checkpoint_v6_writer.go

```diff
@@ -0,0 +1,529 @@
+	}
+	if utilsio.FileExists(fullPath) {
+		return nil, fmt.Errorf("checkpoint file for %v-th sub trie %s already exists", index, fullPath)
+	}
+
+	return createClosableWriter(dir, logger, subTriesFileName, fullPath)
+}
+
+func createClosableWriter(dir string, logger *zerolog.Logger, fileName string, fullPath string) (io.WriteCloser, error) {
+	tmpFile, err := os.CreateTemp(dir, fmt.Sprintf("writing-%v-*", fileName))
+	if err != nil {
+		return nil, fmt.Errorf("could not create temporary file for checkpoint toptries: %w", err)
+	}
+
+	writer := bufio.NewWriterSize(tmpFile, defaultBufioWriteSize)
+	return &SyncOnCloseRenameFile{
+		logger:     logger,
+		file:       tmpFile,
+		targetName: fullPath,
+		Writer:     writer,
+	}, nil
+}
+
+// subtrie file contains:
+// 1. checkpoint version // TODO
+// 2. nodes
+// 3. node count
+// 4. checksum
+func storeCheckpointSubTrie(
+	i int,
+	roots []*node.Node,
+	estimatedSubtrieNodeCount int,
+	outputDir string,
+	outputFile string,
+	logger *zerolog.Logger,
+) (
+	map[*node.Node]uint64, uint64, uint32, error) {
+
+	// traversedSubtrieNodes contains all unique nodes of subtries of the same path and their index.
+	traversedSubtrieNodes := make(map[*node.Node]uint64, estimatedSubtrieNodeCount)
+	// index 0 is nil, it can be used in a node's left child or right child to indicate
+	// a node's left child or right child is nil
+	traversedSubtrieNodes[nil] = 0
+
+	closable, err := createWriterForSubtrie(outputDir, outputFile, logger, i)
+	if err != nil {
+		return nil, 0, 0, fmt.Errorf("could not create writer for sub trie: %w", err)
+	}
+
+	defer func() {
+		closeErr := closable.Close()
+		// Return close error if there isn't any prior error to return.
+		if err == nil {
+			err = closeErr
+		}
+	}()
+
+	// create a CRC32 writer, so that any bytes passed to the writer will
+	// be used to calculate CRC32 checksum
+	writer := NewCRC32Writer(closable)
+
+	// topLevelNodes contains all unique nodes of given tries
+	// from root to subtrie root and their index
+	// (ordered by node traversal sequence).
+	// Index 0 is a special case with nil node.
+	subtrieRootNodes := make(map[*node.Node]uint64, 1<<(subtrieLevel+1))
+	subtrieRootNodes[nil] = 0
+
+	// nodeCounter is counter for all unique nodes.
+	// It starts from 1, as 0 marks nil node.
+	nodeCounter := uint64(1)
+
+	logging := logProgress(fmt.Sprintf("storing %v-th sub trie roots", i), estimatedSubtrieNodeCount, logger)
+	scratch := make([]byte, 1024*4)
+	for _, root := range roots {
+		// Note: nodeCounter is to assign an global index to each node in the order of it being seralized
+		// into the checkpoint file. Therefore, it has to be reused when iterating each subtrie.
+		// storeUniqueNodes will add the unique visited node into traversedSubtrieNodes with key as the node
+		// itself, and value as n-th node being seralized in the checkpoint file.
+		nodeCounter, err = storeUniqueNodes(root, traversedSubtrieNodes, nodeCounter, scratch, writer, logging)
```

Comment:
How compressible is this data?  Can we try lz4 or zstd here to see if there is an upside in compressing this?

Reply:
maybe 50%, it's a trade-off to compress the data, as we have to de-compress it. 

During spork, we used to compress them, but it ended up taking longer than not compressing

Reply:
Interesting!  Lz4 is quite capable at decompression (with C decoder reaching speeds of ~5Gbyte/s).  Did we use [ConcurrencyOption](https://github.com/pierrec/lz4/blob/v4.1.17/options.go#L114)?

Reply:
nevermind the checksumming point-- i do see the benefit of a single end-to-end checksum instead of many per-frame ones.